imagebitmap: Add missing basic functionality

tharkum opened the pull request:
Add missing basic functionality for ImageBitmap
https://html.spec.whatwg.org/multipage/#imagebitmap
including new variant of creation bitmap with source rectangle
https://html.spec.whatwg.org/multipage/#dom-createimagebitmap
but without support of cropping bitmap data with formatting.

Add ImageBitmap to CanvasImageSource union type
https://html.spec.whatwg.org/multipage/#canvasimagesource

Add ImageBitmap to TexImageSource for WebGL
https://registry.khronos.org/webgl/specs/latest/1.0/index.html

Testing: Improvements in the following WPT tests
 - html/canvas/element/manual/imagebitmap/*
 - html/canvas/element/manual/wide-gamut-canvas/*
 - html/semantics/embedded-content/the-canvas-element/*
 - webgl/tests/conformance/textures/image_bitmap_from*
 - webmessaging/postMessage_cross_domain_image_transfer_2d.sub.htm

Fixes: https://github.com/servo/servo/issues/34112

tharkum commented:
@jdm, @sagudev < Please take a look and share your feedback.

I used "Snapshot" as basic brick (CPU-based image memory with metadata) for ImageBitmap and unified behaviour with another HTML elements (e.g. video).

NOTE: WPT results significantly depends from the following PR:  https://github.com/servo/servo/pull/37002 (so please do not pay attention to them right now)

cc @xiaochengh

tharkum commented on components/script/canvas_state.rs:
```diff
@@ -438,6 +452,14 @@ impl CanvasState {
                 self.draw_html_canvas_element(canvas, htmlcanvas, sx, sy, sw, sh, dx, dy, dw, dh)
             },
+            CanvasImageSource::ImageBitmap(ref bitmap) => {
+                // <https://html.spec.whatwg.org/multipage/#check-the-usability-of-the-image-argument>
```
I think link is right https://html.spec.whatwg.org/multipage/canvas.html#check-the-usability-of-the-image-argument

I am checking against `is_detached()` according to specification instead on creation of new function `is_usable()` for ImageBitmap which will do to the same checking `is_detached()` but inside.

I tried to avoid creation of useless wrapper function.

```
- [ImageBitmap](https://html.spec.whatwg.org/multipage/imagebitmap-and-animations.html#imagebitmap)

If image's [[[Detached]]](https://html.spec.whatwg.org/multipage/structured-data.html#detached) internal slot value is set to true, then throw an ["InvalidStateError"](https://webidl.spec.whatwg.org/#invalidstateerror) [DOMException](https://webidl.spec.whatwg.org/#dfn-DOMException).
```

tharkum commented:
+ Added 2 missing WPT results

tharkum commented:
+ Rebased on main branch (changes to pixels::Image)
+ Added derive(MallocSizeOf) for snapshot::Snapshot

sagudev reviewed (commented):
I will just submit my partial review, before it will become to much obsolete, as I will probably not manage to review this before weekend.

tharkum commented on components/pixels/lib.rs:
```diff
@@ -53,6 +67,69 @@ pub fn rgba8_get_rect(pixels: &[u8], size: Size2D<u64>, rect: Rect<u64>) -> Cow<
+    }
+}
+
+pub fn rgba8_flip_y_inplace(pixels: &mut [u8], size: Size2D<u64>) {
+    assert_eq!(pixels.len() % 4, 0);
+
+    let (left, right) = pixels.split_at_mut(pixels.len() / 2);
+
+    let row_length = size.width as usize * 4;
+    let half_height = (size.height / 2) as usize;
+
+    for i in 0..half_height {
+        let top = &mut left[i * row_length..][..row_length];
+        let bottom = &mut right[(half_height - i - 1) * row_length..][..row_length];
+        top.swap_with_slice(bottom);
+    }
+}
```
Sure, I will add some test for flip_y_inplace functionality

sagudev commented on components/script/dom/globalscope.rs:
```diff
@@ -2964,64 +2967,403 @@ impl GlobalScope {
+                    snapshot::AlphaMode::Transparent {
+                        premultiplied: true,
+                    },
+                    snapshot::PixelFormat::BGRA,
```
Ah, I see. Thanks for great image. I still think we should keep original in ImageBitmap, so we will not do any redundant work in WebGL/WebGPU case, but if we are drawing to 2D, we do conversion in ImageBitmap (we change it mutability), so we will conversion only once. But let's do that in follow up.

tharkum committed:
imagebitmap: Add missing basic functionality

Add missing basic functionality for ImageBitmap
https://html.spec.whatwg.org/multipage/#imagebitmap
including new variant of creation bitmap with source rectangle
https://html.spec.whatwg.org/multipage/#dom-createimagebitmap
but without support of cropping bitmap data with formatting.

Add ImageBitmap to CanvasImageSource union type
https://html.spec.whatwg.org/multipage/#canvasimagesource

Add ImageBitmap to TexImageSource for WebGL
https://registry.khronos.org/webgl/specs/latest/1.0/index.html

Testing: Improvements in the following WPT tests
 - html/canvas/element/manual/imagebitmap/*
 - html/canvas/element/manual/wide-gamut-canvas/*
 - html/semantics/embedded-content/the-canvas-element/*
 - webgl/tests/conformance/textures/image_bitmap_from*
 - webmessaging/postMessage_cross_domain_image_transfer_2d.sub.htm

Fixes: #34112

Signed-off-by: Andrei Volykhin <andrei.volykhin@gmail.com>

jdm reviewed (approved):
This is looking good!

tharkum commented on components/script/dom/globalscope.rs:
```diff
@@ -2956,64 +2959,206 @@ impl GlobalScope {
         let in_realm_proof = AlreadyInRealm::assert::<crate::DomTypeHolder>();
         let p = Promise::new_in_current_realm(InRealm::Already(&in_realm_proof), can_gc);
+
+        // Step 1.
```
Done

tharkum commented:
Added missing WPT result for [webmessaging/postMessage_cross_domain_image_transfer_2d.sub.htm](https://github.com/servo/servo/blob/main/tests/wpt/tests/webmessaging/postMessage_cross_domain_image_transfer_2d.sub.htm)

`ctx.drawImage(ImageBitmap) doesn't throw exception anymore (in worker)`
